xarray/core/accessor_str.py

```diff
@@ -865,49 +1061,838 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+        ...             "",
+        ...             "abcdef_Xy_101-fef_Xy_5543210",
+        ...         ],
+        ...     ],
+        ...     dims=["X", "Y"],
+        ... )
+
+        Extract matches
+
+        >>> value.str.extractall(r"(\\w+)_Xy_(\\d*)", group_dim="XX", match_dim="YY")
+        <xarray.DataArray (X: 2, Y: 3, XX: 3, YY: 2)>
+        array([[[['a', '0'],
+                 ['', ''],
+                 ['', '']],
+        <BLANKLINE>
+                [['bab', '110'],
+                 ['baab', '1100'],
+                 ['', '']],
+        <BLANKLINE>
+                [['abc', '01'],
+                 ['cbc', '2210'],
+                 ['', '']]],
+        <BLANKLINE>
+        <BLANKLINE>
+               [[['abcd', ''],
+                 ['dcd', '33210'],
+                 ['dccd', '332210']],
+        <BLANKLINE>
+                [['', ''],
+                 ['', ''],
+                 ['', '']],
+        <BLANKLINE>
+                [['abcdef', '101'],
+                 ['fef', '5543210'],
+                 ['', '']]]], dtype='<U7')
+        Dimensions without coordinates: X, Y, XX, YY
+
+
+        See Also
+        --------
+        DataArray.str.extract
+        DataArray.str.findall
+        re.compile
+        re.findall
+        pandas.Series.str.extractall
+        """
+        pat = self._re_compile(pat, flags, case)
+
+        if pat.groups == 0:
+            raise ValueError("No capture groups found in pattern.")
+
+        if group_dim in self._obj.dims:
+            raise KeyError(f"Group dimension {group_dim} already present in DataArray.")
+
+        if match_dim in self._obj.dims:
+            raise KeyError(f"Match dimension {match_dim} already present in DataArray.")
+
+        if group_dim == match_dim:
+            raise KeyError(
+                f"Group dimension {group_dim} is the same as match dimension {match_dim}."
+            )
+
+        _get_count = lambda x: len(pat.findall(x))
+        maxcount = self._apply(_get_count, dtype=np.int_).max().data.tolist()
+
+        def _get_res(val, pat=pat, num=maxcount, dtype=self._obj.dtype):
+            matches = pat.findall(val)
+            res = np.zeros([maxcount, pat.groups], dtype)
+
+            if pat.groups == 1:
+                for imatch, match in enumerate(matches):
+                    res[imatch, 0] = match
+            else:
+                for imatch, match in enumerate(matches):
+                    for jmatch, submatch in enumerate(match):
+                        res[imatch, jmatch] = submatch
+
+            return res
+
+        return self._apply(
+            # dtype MUST be object or strings can be truncated
+            # See: https://github.com/numpy/numpy/issues/8352
+            _get_res,
+            dtype=np.object_,
+            output_core_dims=[[group_dim, match_dim]],
+            output_sizes={group_dim: pat.groups, match_dim: maxcount},
+        ).astype(self._obj.dtype.kind)
+
+    def findall(
```

Comment:
I am a bit skeptical how useful a DataArray of lists is.

Reply:
It would be a step in further processing I would think.